File: board-service/internal/repository/attachment_repository.go
package repository

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"project-board-api/internal/domain"
)

// AttachmentRepository defines the interface for attachment data access
type AttachmentRepository interface {
	Create(ctx context.Context, attachment *domain.Attachment) error
	FindByID(ctx context.Context, id uuid.UUID) (*domain.Attachment, error)
	FindByEntityID(ctx context.Context, entityType domain.EntityType, entityID uuid.UUID) ([]*domain.Attachment, error)
	FindByIDs(ctx context.Context, ids []uuid.UUID) ([]*domain.Attachment, error)
	Delete(ctx context.Context, id uuid.UUID) error
	FindExpiredTempAttachments(ctx context.Context) ([]*domain.Attachment, error)
	ConfirmAttachments(ctx context.Context, attachmentIDs []uuid.UUID, entityID uuid.UUID) error
	DeleteBatch(ctx context.Context, attachmentIDs []uuid.UUID) error
}

// attachmentRepositoryImpl is the GORM implementation of AttachmentRepository
type attachmentRepositoryImpl struct {
	db *gorm.DB
}

// NewAttachmentRepository creates a new instance of AttachmentRepository
func NewAttachmentRepository(db *gorm.DB) AttachmentRepository {
	return &attachmentRepositoryImpl{db: db}
}

// Create creates a new attachment
func (r *attachmentRepositoryImpl) Create(ctx context.Context, attachment *domain.Attachment) error {
	if err := r.db.WithContext(ctx).Create(attachment).Error; err != nil {
		return err
	}
	return nil
}

// FindByID finds an attachment by its ID
func (r *attachmentRepositoryImpl) FindByID(ctx context.Context, id uuid.UUID) (*domain.Attachment, error) {
	var attachment domain.Attachment
	if err := r.db.WithContext(ctx).First(&attachment, id).Error; err != nil {
		return nil, err
	}
	return &attachment, nil
}

// FindByEntityID finds all attachments by entity type and entity ID
func (r *attachmentRepositoryImpl) FindByEntityID(ctx context.Context, entityType domain.EntityType, entityID uuid.UUID) ([]*domain.Attachment, error) {
	var attachments []*domain.Attachment
	if err := r.db.WithContext(ctx).
		Where("entity_type = ? AND entity_id = ?", entityType, entityID).
		Order("created_at DESC").
		Find(&attachments).Error; err != nil {
		return nil, err
	}
	return attachments, nil
}

// FindByIDs finds attachments by their IDs
func (r *attachmentRepositoryImpl) FindByIDs(ctx context.Context, ids []uuid.UUID) ([]*domain.Attachment, error) {
	if len(ids) == 0 {
		return []*domain.Attachment{}, nil
	}

	var attachments []*domain.Attachment
	if err := r.db.WithContext(ctx).
		Where("id IN ?", ids).
		Find(&attachments).Error; err != nil {
		return nil, err
	}
	return attachments, nil
}

// Delete soft deletes an attachment by ID
func (r *attachmentRepositoryImpl) Delete(ctx context.Context, id uuid.UUID) error {
	if err := r.db.WithContext(ctx).Delete(&domain.Attachment{}, id).Error; err != nil {
		return err
	}
	return nil
}

// FindExpiredTempAttachments finds all temporary attachments that have exceeded their expiration time
func (r *attachmentRepositoryImpl) FindExpiredTempAttachments(ctx context.Context) ([]*domain.Attachment, error) {
	var attachments []*domain.Attachment
	if err := r.db.WithContext(ctx).
		Where("status = ? AND expires_at < ?", domain.AttachmentStatusTemp, time.Now()).
		Find(&attachments).Error; err != nil {
		return nil, err
	}
	return attachments, nil
}

func (r *attachmentRepositoryImpl) ConfirmAttachments(ctx context.Context, attachmentIDs []uuid.UUID, entityID uuid.UUID) error {
	if len(attachmentIDs) == 0 {
		return nil
	}

	// ✅ TEMP 상태만 업데이트, 결과 검증
	result := r.db.WithContext(ctx).
		Model(&domain.Attachment{}).
		Where("id IN ? AND status = ?", attachmentIDs, domain.AttachmentStatusTemp). // ✅
		Updates(map[string]interface{}{
			"status":    domain.AttachmentStatusConfirmed,
			"entity_id": entityID,
		})

	if result.Error != nil {
		return result.Error
	}

	// ✅ 업데이트된 행 수 검증
	if result.RowsAffected == 0 {
		return fmt.Errorf("no attachments were confirmed: all %d attachment(s) are either not found or already confirmed",
			len(attachmentIDs))
	}

	if result.RowsAffected != int64(len(attachmentIDs)) {
		return fmt.Errorf("expected to confirm %d attachment(s) but only confirmed %d",
			len(attachmentIDs), result.RowsAffected)
	}

	return nil
}

// DeleteBatch deletes multiple attachments by their IDs
func (r *attachmentRepositoryImpl) DeleteBatch(ctx context.Context, attachmentIDs []uuid.UUID) error {
	if len(attachmentIDs) == 0 {
		return nil
	}

	if err := r.db.WithContext(ctx).
		Where("id IN ?", attachmentIDs).
		Delete(&domain.Attachment{}).Error; err != nil {
		return err
	}
	return nil
}

File: user-service/src/main/java/OrangeCloud/UserRepo/repository/AttachmentRepository.java
package OrangeCloud.UserRepo.repository;

import OrangeCloud.UserRepo.entity.Attachment;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Attachment 엔티티에 대한 데이터 접근 계층 (DAO)
 */
public interface AttachmentRepository extends JpaRepository<Attachment, UUID> {

    /**
     * Entity ID로 첨부파일 조회
     */
    List<Attachment> findByEntityTypeAndEntityIdAndDeletedAtIsNull(
            Attachment.EntityType entityType, 
            UUID entityId);

    /**
     * 만료된 임시 첨부파일 조회
     * status가 TEMP이고 expiresAt이 현재 시간보다 이전인 파일들
     */
    @Query("SELECT a FROM Attachment a WHERE a.status = 'TEMP' " +
           "AND a.expiresAt < :now AND a.deletedAt IS NULL")
    List<Attachment> findExpiredTempAttachments(@Param("now") LocalDateTime now);

    /**
     * ID로 첨부파일 조회 (삭제되지 않은 것만)
     */
    Optional<Attachment> findByIdAndDeletedAtIsNull(UUID id);

    /**
     * 여러 ID로 첨부파일 조회 (삭제되지 않은 것만)
     */
    List<Attachment> findByIdInAndDeletedAtIsNull(List<UUID> ids);

    /**
     * 첨부파일 확정 (TEMP -> CONFIRMED)
     */
    @Modifying
    @Query("UPDATE Attachment a SET a.status = 'CONFIRMED', a.entityId = :entityId, " +
           "a.expiresAt = NULL WHERE a.id IN :attachmentIds")
    int confirmAttachments(@Param("attachmentIds") List<UUID> attachmentIds, 
                          @Param("entityId") UUID entityId);

    /**
     * 여러 첨부파일 일괄 삭제 (soft delete)
     */
    @Modifying
    @Query("UPDATE Attachment a SET a.deletedAt = :deletedAt WHERE a.id IN :attachmentIds")
    int softDeleteBatch(@Param("attachmentIds") List<UUID> attachmentIds, 
                       @Param("deletedAt") LocalDateTime deletedAt);

    /**
     * 여러 첨부파일 일괄 삭제 (hard delete)
     */
    @Modifying
    @Query("DELETE FROM Attachment a WHERE a.id IN :attachmentIds")
    int deleteBatch(@Param("attachmentIds") List<UUID> attachmentIds);
}
